cmd/samples/pso/workflow.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"go.uber.org/cadence"
+	"go.uber.org/zap"
+
+	"github.com/pborman/uuid"
+	"go.uber.org/cadence/workflow"
+)
+
+type (
+	individual struct {
+		IndividualID string
+		Variable     []float64
+		Fitness      []float64
+	}
+)
+
+// ApplicationName is the task list for this sample
+const ApplicationName = "PSO"
+
+// HostID - Use a new uuid just for demo so we can run 2 host specific activity workers on same machine.
+// In real world case, you would use a hostname or ip address as HostID.
+var HostID = ApplicationName + "_" + uuid.New()
+
+var ActivityOptions = workflow.ActivityOptions{
+	ScheduleToStartTimeout: time.Second * 5,
+	StartToCloseTimeout:    time.Minute * 10,
+	HeartbeatTimeout:       time.Second * 2, // such a short timeout to make sample fail over very fast
+	RetryPolicy: &cadence.RetryPolicy{
+		InitialInterval:          time.Second,
+		BackoffCoefficient:       2.0,
+		MaximumInterval:          time.Minute,
+		ExpirationInterval:       time.Minute * 10,
+		MaximumAttempts:          5,
+		NonRetriableErrorReasons: []string{"bad-error"},
+	},
+}
+
+var QueryResult string
+
+// This is registration process where you register all your workflow handlers.
+func init() {
+	//workflow.Register(FitnessEvaluationWorkflow)
+	workflow.Register(PSOWorkflow)
+	workflow.Register(PSOChildWorkflow)
+}
+
+//PSOWorkflow workflow decider
+func PSOWorkflow(ctx workflow.Context, functionName string) (err error) {
+	logger := workflow.GetLogger(ctx)
+
+	// Set activity options
+	ctx = workflow.WithActivityOptions(ctx, ActivityOptions)
+
+	// Set child workflow options
+	execution := workflow.GetInfo(ctx).WorkflowExecution
+	// Parent workflow can choose to specify it's own ID for child execution.  Make sure they are unique for each execution.
+	childID := fmt.Sprintf("PSO_child_workflow:%v", execution.RunID)
+	cwo := workflow.ChildWorkflowOptions{
+		// Do not specify WorkflowID if you want cadence to generate a unique ID for child execution
+		WorkflowID:                   childID,
+		ExecutionStartToCloseTimeout: time.Minute,
+	}
+	ctx = workflow.WithChildOptions(ctx, cwo)
+
+	// setup query handler for query type "state"
+	QueryResult = "started"
```

Comment:
Cannot use global variables to share data across multipel workflow instances. The solution is to have two queries. One in the parent workflow to return the id of the currently running child the other one is in the child to return its current status.
It currently works only because a single worker process is used. As soon as this workflow is deployed to a cluster the query is going to break.

Reply:
Ok, maybe I got it, see my comment above too